core/indigo-core/reaction/src/reaction_multistep_detector.cpp

```diff
@@ -435,41 +950,153 @@ bool ReactionMultistepDetector::mapMultitailReactionComponents()
+            auto& mdj = undef_distances[j];
+            if (dist < LayoutOptions::DEFAULT_BOND_LENGTH * 2)
+            {
+                mdi.sorted_distances.emplace_back(j, dist);
+                mdj.sorted_distances.emplace_back(i, dist);
+            }
+            mdj.distances_map.emplace(i, dist);
+            mdi.distances_map.emplace(j, dist);
         }
+        std::sort(mdi.sorted_distances.begin(), mdi.sorted_distances.end(), [](auto& lhs, auto& rhs) { return lhs.second < rhs.second; });
+
+        // collect distances to detected components
+        auto& mdic = undef_comp_distances[i];
+        for (size_t j = 0; j < component_bboxes.size(); ++j)
+        {
+            float dist = undef_component_bboxes[i].second.distanceTo(component_bboxes[j].second);
+            mdic.distances_map.emplace(j, dist);
+            if (dist < LayoutOptions::DEFAULT_BOND_LENGTH * 2)
+                mdic.sorted_distances.emplace_back(j, dist);
+        }
+        std::sort(mdic.sorted_distances.begin(), mdic.sorted_distances.end(), [](auto& lhs, auto& rhs) { return lhs.second < rhs.second; });
     }
-    return !bad_pathway;
+}
+
+void ReactionMultistepDetector::mergeUndefinedComponents()
+{
+    std::vector<std::pair<size_t, Rect2f>> component_bboxes;
+    std::vector<std::pair<size_t, Rect2f>> undef_component_bboxes;
+    MOL_DISTANCES mol_distances;
+
+    for (size_t i = 0; i < _reaction_components.size(); ++i)
+    {
+        auto& rc = _reaction_components[i];
+        if (rc.component_type == ReactionComponent::MOLECULE)
+        {
+            auto& csb = _component_summ_blocks[rc.summ_block_idx];
+            Rect2f bbox;
+            rc.molecule->getBoundingBox(bbox);
+            if (csb.role == BaseReaction::UNDEFINED)
+                undef_component_bboxes.emplace_back(i, bbox);
+            else
+                component_bboxes.emplace_back(i, bbox);
+        }
+    }
+
+    bool has_merges = false;
+    do
+    {
+        std::vector<MOL_DISTANCES_DESC> undef_distances;
+        std::vector<MOL_DISTANCES_DESC> undef_comp_distances;
+        collectUndefinedDistances(component_bboxes, undef_component_bboxes, undef_distances, undef_comp_distances);
+
+        // find minimal distance to other undefined components
+        std::optional<MERGE_CANDIDATE> undef_min_dist;
+        std::vector<MERGE_CANDIDATE> merge_list_candidates;
+
+        // iterate over all undefined components
+        for (size_t i = 0; i < undef_component_bboxes.size(); ++i)
+        {
+            auto& mdi_comp = undef_comp_distances[i];
+            auto& mdi_undef = undef_distances[i];
+            // check if we have closest detected component
+            if (mdi_comp.sorted_distances.size())
+            {
+                // check if we have a merge candidate
+                if (mdi_undef.sorted_distances.empty() || mdi_comp.sorted_distances.front().second < mdi_undef.sorted_distances.front().second)
```

Comment:
dont see where sorted distances recalculated

Reply:
it happens in collectUndefinedDistances